azure-spring-boot/src/main/java/com/microsoft/azure/keyvault/spring/KeyVaultOperation.java

```diff
@@ -3,102 +3,15 @@
-    }
-
-    public KeyVaultOperation(final KeyVaultClient keyVaultClient, String vaultUri, final long refreshInterval) {
-        this.cacheRefreshIntervalInMs = refreshInterval;
-        this.keyVaultClient = keyVaultClient;
-
-        vaultUri = vaultUri.trim();
-        if (vaultUri.endsWith("/")) {
-            vaultUri = vaultUri.substring(0, vaultUri.length() - 1);
-        }
-        this.vaultUri = vaultUri;
-
-        createOrUpdateHashMap();
-    }
-
-    public String[] list() {
-        try {
-            this.rwLock.readLock().lock();
-            return Collections.list(this.propertyNamesHashMap.keys())
-                    .toArray(new String[this.propertyNamesHashMap.size()]);
-        } finally {
-            this.rwLock.readLock().unlock();
-        }
-    }
-
-    public Object get(final String secretName) {
-        // NOTE: azure keyvault secret name convention: ^[0-9a-zA-Z-]+$ "." is not allowed
-        final String localSecretName = secretName.replace(".", "-");
-
-        // refresh periodically
-        if (System.currentTimeMillis() - this.lastUpdateTime.get() > this.cacheRefreshIntervalInMs) {
-            synchronized (this.refreshLock) {
-                if (System.currentTimeMillis() - this.lastUpdateTime.get() > this.cacheRefreshIntervalInMs) {
-                    this.lastUpdateTime.set(System.currentTimeMillis());
-                    createOrUpdateHashMap();
-                }
-            }
-        }
-
-        if (this.propertyNamesHashMap.containsKey(secretName)) {
-            final SecretBundle secretBundle = this.keyVaultClient.getSecret(this.vaultUri, localSecretName);
-            return secretBundle.value();
-
-        } else {
-            return null;
-        }
-    }
-
-    private void createOrUpdateHashMap() {
-        if (this.propertyNamesHashMap == null) {
-            this.propertyNamesHashMap = new ConcurrentHashMap<>();
-        }
-
-        try {
-            this.rwLock.writeLock().lock();
-            this.propertyNamesHashMap.clear();
-
-            final PagedList<SecretItem> secrets = this.keyVaultClient.listSecrets(this.vaultUri);
-            secrets.loadAll();
+/**
+ * Azure Key Vault related operation
+ */
+public interface KeyVaultOperation {
+    void setUseCache(boolean useCache);
 
-            for (final SecretItem secret : secrets) {
-                this.propertyNamesHashMap.putIfAbsent(secret.id()
-                        .replaceFirst(this.vaultUri + "/secrets/", "")
-                        .replaceAll("-", "."), secret.id());
-                this.propertyNamesHashMap.putIfAbsent(secret.id()
-                        .replaceFirst(this.vaultUri + "/secrets/", ""), secret.id());
-            }
+    void setRefreshInterval(long refreshIntervalMilliSecs);
 
-            this.lastUpdateTime.set(System.currentTimeMillis());
-        } finally {
-            this.rwLock.writeLock().unlock();
-        }
-    }
+    void setAllowTelemetry(boolean allowTelemetry);
```

Comment:
This shouldn't be part of interface. If others want to provide another implementation. Why should we bother with this?

Reply:
Also for other 2 methods.

Reply:
Moved to abstract template and now only keep the KeyVaultOperation, and we can provide more interfaces if we decide to support certificates etc. in KeyVault.

Reply:
Moved to the abstract template.